sound/soc/sof/topology.c

```diff
@@ -1498,7 +1498,8 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 	case snd_soc_dapm_kcontrol:
 	default:
 		dev_dbg(scomp->dev, "widget type %d name %s not handled\n", swidget->id, tw->name);
-		break;
+		kfree(swidget);
+		return ret;
```

Comment:
What are the virtual widgets' type? Do we need a `break;` before `default`?

Reply:
There are some return without `kfree(swidget);` before `switch (w->id)`. For example, line 1406, 1413, 1422, 1432. Should we use this opportunity to revise the error handling?